Realm/RLMLogger.mm

```diff
@@ -60,90 +58,203 @@ static RLMLogLevel logLevelForLevel(Level logLevel) {
+        }
+        REALM_UNREACHABLE();
+    };
+
+#if REALM_DEBUG
+    // Validate that all of core's categories are present in the map
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        for (auto name : LogCategory::get_category_names()) {
+            (void)find(LogCategory::get_category(name));
+        }
+    });
+#endif
+
+    return find(category);
 }
 
-struct CocoaLogger : public Logger {
+struct DynamicLogger : Logger {
+    RLMUnfairMutex _mutex;
+    NSArray<RLMLogCategoryFunction> *_logFunctions;
+
     void do_log(const LogCategory& category, Level level, const std::string& message) override {
-        NSLog(@"%@:%@ %@", levelPrefix(level), RLMStringDataToNSString(category.get_name()), RLMStringDataToNSString(message));
+        NSArray *loggers;
+        {
+            std::lock_guard lock(_mutex);
+            loggers = _logFunctions;
+        }
+        if (loggers.count == 0) {
+            return;
+        }
+
+        @autoreleasepool {
+            NSString *nsMessage = RLMStringDataToNSString(message);
+            RLMLogCategory rlmCategory = logCategoryForCategory(category);
+            RLMLogLevel rlmLevel = logLevelForLevel(level);
+            for (RLMLogCategoryFunction fn : loggers) {
+                fn(rlmLevel, rlmCategory, nsMessage);
+            }
+        }
     }
 };
+std::optional<DynamicLogger> s_dynamic_logger;
 
 class CustomLogger : public Logger {
 public:
-    RLMLoggerFunction function;
-    void do_log(const LogCategory& category, Level level, const std::string& message) override {
+    RLMLogFunction function;
+    void do_log(const LogCategory&, Level level, const std::string& message) override {
         @autoreleasepool {
-            if (function) {
-                function(logLevelForLevel(level), logCategoryForCategoryName(category.get_name()), RLMStringDataToNSString(message));
-            }
+            function(logLevelForLevel(level), RLMStringDataToNSString(message));
         }
     }
 };
 } // anonymous namespace
 
+@implementation RLMLoggerToken {
+    RLMLogCategoryFunction _function;
+}
+
+- (instancetype)initWithFunction:(RLMLogCategoryFunction)function {
+    if (self = [super init]) {
+        _function = function;
+    }
+    return self;
+}
+
+- (void)invalidate {
+    std::lock_guard lock(s_dynamic_logger->_mutex);
+    if (!_function) {
+        return;
+    }
+    auto& functions = s_dynamic_logger->_logFunctions;
+    functions = [functions filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"SELF != %@", _function]];
+    _function = nil;
+}
+
+@end
+
 @implementation RLMLogger {
     std::shared_ptr<Logger> _logger;
 }
 
-typedef void(^LoggerBlock)(RLMLogLevel level, NSString *message);
+#pragma mark - Dynamic category-based API
+
++ (void)initialize {
+    // Constructing the global logger is deferred slightly because constructing
+    // a Logger references non-constinit global variables (the logger categories)
```

Comment:
The logger categories are global variables and Logger's constructor reads from them. If the dynamic logger is also a global variable, it may get constructed before the categories are constructed.